### PR TITLE
add last part Todo API

### PR DESCRIPTION
**Pull request by AndriiYelieva**
[DEMO LINK](https://AndriiYelieva.github.io/react_todo-app-with-api/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,24 +1,237 @@
...
+        break;
+      case FilterValue.Completed: visibleTodos = visibleTodos.filter(
+        (todo) => todo.completed,
+      );
+        break;
+      default:
+        break;
+    }
+
+    return visibleTodos;
```
Comment (VitaliyBondarenko1982) [inline src/App.tsx:64]:
```suggestion
    switch (filterValue) {
      case FilterValue.Active: 
        return  todos.filter( (todo) => !todo.completed );
      case FilterValue.Completed: 
        return todos.filter( (todo) => todo.completed);
      default:
        return todos;
    }
```
return from switch filtered todos or all todos in default case.

File: src/App.tsx [hunk 2]
```diff
@@ -1,24 +1,237 @@
...
+
+  useEffect(() => {
+    loadTodos();
+  }, []);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const getFilteringTodos = () => {
```
Comment (VitaliyBondarenko1982) [inline src/App.tsx:48]:
```suggestion
  const filteredTodos = useMemo(() => {
```
wrap function in useMemo. it wiil return filteredTodos instead call function in each places where you need actual todos.

File: src/components/Footer.tsx [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  setFilterValue, filterValue, todos, clearCompletedTodos, completedTodos,
+}) => {
+  const activeTodos = todos.filter(todo => !todo.completed);
+
+  return (
+    <footer className="todoapp__footer">
+      <span className="todo-count">
+        {activeTodos.length}
+        {' '}
+        items left
```
Comment (VitaliyBondarenko1982) [inline src/components/Footer.tsx:24]: use template string here
```suggestion
        {`${activeTodos.length} items left`}
```
check other similar cases

File: src/components/TodoFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+export const TodoFilter = ({ setFilterValue, filterValue }: Props) => {
+  return (
+    // eslint-disable-next-line react/jsx-filename-extension
+    <nav className="filter">
+      <a
+        href="#/"
+        className={classNames('filter__link', {
+          selected: filterValue === FilterValue.All,
+        })}
+        onClick={() => setFilterValue(FilterValue.All)}
```
Comment (VitaliyBondarenko1982) [inline src/components/TodoFilter.tsx:19]: make one function factory in component which will be accept filter value and return arrow function.
call it in each link with proper value.
```
const handleFilter = (filterValue: FilterValue) => () => serFilterValue(filterValue)
```
```suggestion
        onClick={handleFilter(FilterValue.All)}
```

**Review by VitaliyBondarenko1982 — CHANGES_REQUESTED** [review]
Please check comments and demo link
Screenshot from 2023-08-16 13-43-21

**Review by IvanFesenko — CHANGES_REQUESTED** [review]
Your app works incorrect
https://github.com/mate-academy/react_todo-app-with-api/assets/61904995/e057b184-3bf6-433e-9d7e-b23ee91bd472

**Review by etojeDenys — CHANGES_REQUESTED** [review]
looks broken :(

https://github.com/mate-academy/react_todo-app-with-api/assets/70220667/854c6824-f21e-450e-8a55-2f86960cc5d8

File: src/Components/TodoFooter.tsx [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        </a>
+      </nav>
+
+      <button
+        type="button"
+        className={classNames('todoapp__clear-completed',
+          { 'is-invisible': completedTodoCount === 0 })}
+        onClick={() => {
+          deleteCompletedTodos();
+        }}
```
Comment (VitaliyBondarenko1982) [inline src/Components/TodoFooter.tsx:96]:
```suggestion
        onClick={deleteCompletedTodos}
```

File: src/Components/TodoHeader.tsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+      }))
+        .then(() => {
+          getTodos(USER_ID)
+            .then((value) => {
+              setTodos(value);
+              setLoadingIds([]);
+            })
+            .catch(() => {
+              setErrorMessage(Notification.load);
+            });
```
Comment (VitaliyBondarenko1982) [inline src/Components/TodoHeader.tsx:73]: it is redundant to load todos after finish update.
we can avoid this request to server. 
you can just update todos state.

File: src/Components/TodoHeader.tsx [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    setLoadingIds(currIds => [...currIds, ...uncompletedTodosIds]);
+
+    Promise.all(uncompletedTodos.map(todo => {
+      return updateTodo(todo.id, { completed: !todo.completed });
+    }))
+      .then(() => {
+        getTodos(USER_ID)
+          .then((value) => {
+            setTodos(value);
+          })
```
Comment (VitaliyBondarenko1982) [inline src/Components/TodoHeader.tsx:91]: same. suggest do not make request to server and just update locally todos.
check other similar cases

File: src/Components/TodoItem.tsx [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+          onSubmit={(event) => handleClickedFormSubmit(
+            event,
+            editingTitle,
+            todo.id,
+          )}
+          onBlur={(event) => handleClickedFormSubmit(
+            event,
+            editingTitle,
+            todo.id,
+          )}
```
Comment (VitaliyBondarenko1982) [inline src/Components/TodoItem.tsx:163]: make separate funcitons for all handlers and jsx and pass reference here

File: src/Components/TodoItem.tsx [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+        />
+      </label>
+
+      {doubleClicked ? (
+        <form
+          onSubmit={(event) => handleClickedFormSubmit(
+            event,
+            editingTitle,
+            todo.id,
+          )}
```
Comment (VitaliyBondarenko1982) [inline src/Components/TodoItem.tsx:158]: make separate funcitons for all handlers and jsx and pass reference here
for example here something like that
check other similar cases

```
const handleClieckedFormSubmit = (id: string) => (event: React.FormEvent<HTMLFormElement>) => {
    // do all logic for submit
}
```

```suggestion
          onSubmit={handleClickedFormSubmit(todo.id)}
```

File: src/Components/TodoFooter.tsx [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        {`${uncompletedTodoCount} items left`}
+      </span>
+
+      <nav className="filter">
+        <a
+          href="#/"
+          className={classNames('filter__link', {
+            selected: filterType === Filter.ALL,
+          })}
+          onClick={() => setFilterType(Filter.ALL)}
```
Comment (VitaliyBondarenko1982) [inline src/Components/TodoFooter.tsx:64]: please fix this comment
Screenshot from 2023-08-17 12-41-13

**Review by VitaliyBondarenko1982 — CHANGES_REQUESTED** [review]
Needs some improvements

**Pull request closed without merging**